### PR TITLE
Nucleo IO board firmware

### PR DESCRIPTION
**Pull request by balazsracz**
Firmware for the new openlcb Dev Kit IO board.
@szabodabo if you could review this that would be great.

### CODE REVIEW

File: src/freertos_drivers/common/MmapGPIO.hxx [hunk 1]
```diff
@@ -0,0 +1,113 @@
+/** \copyright
```
Comment (bakerstu) [inline src/freertos_drivers/common/MmapGPIO.hxx]: I'm not sure that this file belongs here.  Why does not seem specific to FreeRTOS.  Could it not also be used in Linux for example?

File: src/freertos_drivers/st/Stm32SPI.cxx [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+{
+    spi_reset(port);
+    memset(&spiHandle_, 0, sizeof(spiHandle_));
+
+    spiHandle_.Instance = port;
+
+    HASSERT(update_configuration() == 0);
+}
+
+static const uint32_t baud_rate_table[] = {
```
Comment (bakerstu) [inline src/freertos_drivers/st/Stm32SPI.cxx]: sytle:  need hanging brace.

File: src/freertos_drivers/st/Stm32SPI.cxx [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    // Deinit if needed.
+    if (spiHandle_.State != HAL_SPI_STATE_RESET)
+    {
+        HAL_SPI_DeInit(&spiHandle_);
+    }
+
+    // Computes the lowest divisor that gets us under the desired max speed Hz.
+    uint32_t pclock = 0; //cm3_cpu_clock_hz;  //HAL_RCC_GetPCLK1Freq();
+    /// @todo this might not be the correct clock value for the F303
+//#ifdef __LL_RCC_CALC_PCLK1_FREQ
```
Comment (bakerstu) [inline src/freertos_drivers/st/Stm32SPI.cxx]: is the #ifdef, #else dead code?

File: src/openlcb/MultiConfiguredConsumer.hxx [hunk 4]
```diff
@@ -61,7 +62,7 @@ public:
         , size_(N)
         , offset_(config)
     {
-        HASSERT(size_ == N);
+        HASSERT(size == N);
```
Comment (bakerstu) [inline src/openlcb/MultiConfiguredConsumer.hxx]: if size is always == N, should we deprecate this constructor and create another one without the size parameter at all?

File: src/freertos_drivers/common/MmapGPIO.hxx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        : ptr_(ptr + (bit_ofs >> 5))
+        , bit_(bit_ofs & 31)
+        , isOutput_(is_output ? 1 : 0)
+    {
+    }
+
+    void write(Value new_state) const OVERRIDE
+    {
+        if (new_state == Value::SET) {
+            *ptr_ |= (1<<bit_);
```
Comment (szabodabo) [inline src/freertos_drivers/common/MmapGPIO.hxx]: why not just call set() and clr() here?

File: src/openlcb/MultiConfiguredConsumer.hxx [hunk 6]
```diff
@@ -102,6 +103,28 @@ public:
     /// @todo(balazs.racz): implement
```
Comment (szabodabo) [inline src/openlcb/MultiConfiguredConsumer.hxx]: todone?

Comment (balazsracz) [reply src/openlcb/MultiConfiguredConsumer.hxx]: done

File: src/freertos_drivers/common/MmapGPIO.hxx (hunk 5 above)
Comment (balazsracz) [reply src/freertos_drivers/common/MmapGPIO.hxx]: those are virtual functions so if I call them here we will have two virtual trampolines in the execution cost or each write call.

File: boards/st-stm32f303re-nucleo/HwInit.cxx [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file HwInit.cxx
+ * This file represents the hardware initialization for the TI Tiva MCU.
```
[the remark this answers is not in the document]
Comment (balazsracz) [reply boards/st-stm32f303re-nucleo/HwInit.cxx]: done. Fixed for the 091 too.

File: boards/st-stm32f303re-nucleo-dev-board/HwInit.cxx [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file HwInit.cxx
+ * This file represents the hardware initialization for the TI Tiva MCU.
```
[the remark this answers is not in the document]
Comment (balazsracz) [reply boards/st-stm32f303re-nucleo-dev-board/HwInit.cxx]: done

File: applications/nucleo_io/targets/freertos.armv6m.st-stm32f091rc-nucleo-dev-board/main.cxx [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file main.cxx
+ *
+ * Main file for the io board application on the Tiva Launchpad board.
+ *
```
[the remark this answers is not in the document]
Comment (balazsracz) [reply applications/nucleo_io/targets/freertos.armv6m.st-stm32f091rc-nucleo-dev-board/main.cxx]: done

File: applications/nucleo_io/targets/freertos.armv6m.st-stm32f091rc-nucleo-dev-board/main.cxx [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+const unsigned servo_max = configCPU_CLOCK_HZ * 2 / 1000;
+
+PWMGPO srv1_gpo(servo_channels[0], servo_min, servo_max);
+PWMGPO srv2_gpo(servo_channels[1], servo_min, servo_max);
+PWMGPO srv3_gpo(servo_channels[2], servo_min, servo_max);
+PWMGPO srv4_gpo(servo_channels[3], servo_min, servo_max);
+
+constexpr const Gpio *const kServoGpio[] = {
+    &srv1_gpo, &srv2_gpo, &srv3_gpo, &srv4_gpo, //
+    SRV5_Pin::instance(), SRV6_Pin::instance(), //
```
[the remark this answers is not in the document]
Comment (balazsracz) [reply applications/nucleo_io/targets/freertos.armv6m.st-stm32f091rc-nucleo-dev-board/main.cxx]: TL;DR: 1-4 works, 5-8 does not :)
1-4 have hardware PWM channels with a dedicated timer. The waveform gets generated by the hardware timer's PWM output, so all we need to do is configure the hardware timer correctly. The STM32PWM.hxx driver will do that.
5-8 are just GPIO pins and for the moment they only generate a 0 or 1 constant voltge level. That's not super useful for a servo, I agree. Someone needs to write a software PWM library that will generate the servo output waveform for these before they can be used to drive servos. That's not super difficult; there are plenty of hardware timer resources available to write an interrupt-driven PWM generator.

File: applications/io_board/targets/freertos.armv7m.st-stm32f303re-nucleo/main.cxx [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+// layout. The argument of offset zero is ignored and will be removed later.
+openlcb::ConfigDef cfg(0);
+// Defines weak constants used by the stack to tell it which device contains
+// the volatile configuration information. This device name appears in
+// HwInit.cxx that creates the device drivers.
+extern const char *const openlcb::CONFIG_FILENAME = "/dev/eeprom";
+// The size of the memory space to export over the above device.
+extern const size_t openlcb::CONFIG_FILE_SIZE =
+    cfg.seg().size() + cfg.seg().offset();
+static_assert(openlcb::CONFIG_FILE_SIZE <= 300, "Need to adjust eeprom size");
```
[the remark this answers is not in the document]
Comment (balazsracz) [reply applications/io_board/targets/freertos.armv7m.st-stm32f303re-nucleo/main.cxx]: Those are different applications. The bare io-board application only has one button and one LED, accordingly the configuration is pretty small so we don't need a lot of (fake) eeprom for it.
The nucleo-dev-board has a lot of IOs and accordingly a pretty big configuration "file", so the assertion is different too.

File: applications/io_board/targets/freertos.armv7m.st-stm32f303re-nucleo/config.hxx [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+/// - Hardware version
+/// - Software version
+///
+/// This data will be used for all purposes of the identification:
+///
+/// - the generated cdi.xml will include this data
+/// - the Simple Node Ident Info Protocol will return this data
+/// - the ACDI memory space will contain this data.
+extern const SimpleNodeStaticValues SNIP_STATIC_DATA = {
+    4,               "OpenMRN", "Test IO Board - STM32F091RC Nucleo",
```
[the remark this answers is not in the document]
Comment (balazsracz) [reply applications/io_board/targets/freertos.armv7m.st-stm32f303re-nucleo/config.hxx]: done

File: src/openlcb/MultiConfiguredConsumer.hxx (hunk 4 above)
Comment (balazsracz) [reply src/openlcb/MultiConfiguredConsumer.hxx]: This is a bit of a template trickery to ensure that two arrays are of the same length without the user having to specify either of them. A typical usage is this:
```
openlcb::MultiConfiguredConsumer direct_consumers(stack.node(), kDirectGpio,
    ARRAYSIZE(kDirectGpio), cfg.seg().direct_consumers());
```
Here both N and size are computed by the compiler. Unfortunately we cannot check them statically (without resorting to some ugly macro), but the device will crash deterministically upon first boot if they do not match.

I added documentation comments on this.

File: src/freertos_drivers/st/Stm32SPI.cxx (hunk 3 above)
Comment (balazsracz) [reply src/freertos_drivers/st/Stm32SPI.cxx]: fixed

File: src/freertos_drivers/st/Stm32SPI.cxx (hunk 2 above)
Comment (balazsracz) [reply src/freertos_drivers/st/Stm32SPI.cxx]: done

**Pull request merged**